xwayland/selection/incoming.c

```diff
@@ -25,113 +40,87 @@ static int xwm_data_source_write(int fd, uint32_t mask, void *data) {
+	} else {
+		wlr_log(WLR_DEBUG, "transfer complete");
+		xwm_selection_transfer_finish(transfer);
 	}
 
-	return 1;
+	return 0;
 }
 
-static void xwm_write_property(struct wlr_xwm_selection_transfer *transfer,
-		xcb_get_property_reply_t *reply) {
-	struct wlr_xwm *xwm = transfer->selection->xwm;
-
-	transfer->property_start = 0;
-	transfer->property_reply = reply;
-
-	xwm_data_source_write(transfer->wl_client_fd, WL_EVENT_WRITABLE, transfer);
+static void xwm_write_selection_property_to_wl_client(
+		struct wlr_xwm_selection_transfer *transfer) {
+	if (transfer->incr && transfer->wl_client_fd < 0) {
+		// Wayland client closed its pipe prematurely before the X11 client finished
+		// its incremental transfer. Continue draining the X11 client.
+		xwm_notify_ready_for_next_incr_chunk(transfer);
+		return;
+	}
 
-	if (transfer->property_reply != NULL) {
+	bool wl_client_finished_consuming = !write_selection_property_to_wl_client(
+			transfer->wl_client_fd, WL_EVENT_WRITABLE, transfer);
+	if (!wl_client_finished_consuming) {
+		// Wrote out part of the property to the Wayland client, but the client was
+		// unable to accept all of it. Schedule an event to asynchronously complete
+		// the transfer.
 		struct wl_event_loop *loop =
-			wl_display_get_event_loop(xwm->xwayland->wl_display);
+			wl_display_get_event_loop(transfer->selection->xwm->xwayland->wl_display);
 		transfer->event_source = wl_event_loop_add_fd(loop,
-			transfer->wl_client_fd, WL_EVENT_WRITABLE, xwm_data_source_write,
-			transfer);
+			transfer->wl_client_fd, WL_EVENT_WRITABLE,
+			write_selection_property_to_wl_client, transfer);
 	}
 }
 
 void xwm_get_incr_chunk(struct wlr_xwm_selection_transfer *transfer) {
-	struct wlr_xwm *xwm = transfer->selection->xwm;
 	wlr_log(WLR_DEBUG, "xwm_get_incr_chunk");
 
-	xcb_get_property_cookie_t cookie = xcb_get_property(xwm->xcb_conn,
-		0, // delete
-		transfer->selection->window,
-		xwm->atoms[WL_SELECTION],
-		XCB_GET_PROPERTY_TYPE_ANY,
-		0, // offset
-		0x1fffffff // length
-		);
+	if (transfer->property_reply) {
+		wlr_log(WLR_ERROR, "X11 client offered a new property before we deleted");
+		return;
+	}
 
-	xcb_get_property_reply_t *reply =
-		xcb_get_property_reply(xwm->xcb_conn, cookie, NULL);
-	if (reply == NULL) {
-		wlr_log(WLR_ERROR, "cannot get selection property");
+	if (!xwm_selection_transfer_get_selection_property(transfer, false)) {
 		return;
 	}
-	//dump_property(xwm, xwm->atoms[WL_SELECTION], reply);
 
-	if (xcb_get_property_value_length(reply) > 0) {
-		/* Reply's ownership is transferred to xwm, which is responsible
-		 * for freeing it */
-		xwm_write_property(transfer, reply);
+	if (xcb_get_property_value_length(transfer->property_reply) > 0) {
+		xwm_write_selection_property_to_wl_client(transfer);
 	} else {
-		wlr_log(WLR_DEBUG, "transfer complete");
-		xwm_selection_transfer_close_wl_client_fd(transfer);
-		free(reply);
+		wlr_log(WLR_DEBUG, "incremental transfer complete");
+		transfer->incr = false;
```

Comment:
Maybe we want this to be part of `xwm_selection_transfer_finish`. I haven't thought too much about it.

Reply:
Yeah, I think I like that more.